cmd/analyze_deps.go

```diff
@@ -30,99 +27,30 @@ to quickly create a Cobra application.`,
 			return err
 		}
 
-		// get output of "go mod graph" in a string
-		goModGraph := exec.Command("go", "mod", "graph")
-		goModGraphOutput, err := goModGraph.Output()
-		if err != nil {
-			return err
-		}
-		goModGraphOutputString := string(goModGraphOutput)
-
-		// create a graph of dependencies from that output
-		depGraph := make(map[string][]string)
-		scanner := bufio.NewScanner(strings.NewReader(goModGraphOutputString))
-
-		// deps will store all the dependencies
-		// since can't do slice.contains so better to use map
-		var deps []string
-		mainModule := "notset"
-
-		for scanner.Scan() {
-			line := scanner.Text()
-			words := strings.Fields(line)
-			// remove versions
-			words[0] = (strings.Split(words[0], "@"))[0]
-			words[1] = (strings.Split(words[1], "@"))[0]
-			if mainModule == "notset" {
-				mainModule = words[0]
-			}
-			if !contains(deps, words[0]) {
-				deps = append(deps, words[0])
-			}
-			if !contains(deps, words[1]) {
-				deps = append(deps, words[1])
-			}
-
-			// we don't want to add the same dep again
-			if !contains(depGraph[words[0]], words[1]) {
-				depGraph[words[0]] = append(depGraph[words[0]], words[1])
-			}
-		}
+		depGraph, deps, mainModule := getDepInfo()
 
 		if verbose {
 			fmt.Println("All dependencies:")
-			for _, v := range deps {
-				if v == mainModule {
-					continue
-				}
-				fmt.Println(v)
-			}
-			fmt.Println()
+			printDeps(deps)
 		}
 
-		// Prepare Dynamic Programming arrays for max depth
-		// dp[k] = max depth of dependencies starting from dependency "k"
-		dp := make(map[string]int)
-		// visited array will make sure we don't have infinite recursion
-		visited := make(map[string]bool)
-		recVisited := make(map[string]bool)
-
-		// values not in map will have their respective 0 value by default
-		// so need to worry about terminal nodes
-		for _, v := range deps {
-			dp[v] = 0
-			visited[v] = false
-			recVisited[v] = false
-		}
-		// longestPath[k] = u means that from dependency "k" going to
-		// dependency "u" will result in the longest path
-		longestPath := make(map[string]string)
-
-		// maps are pass by reference in golang
-		// longest path would always start from the main module
-		dfs(mainModule, depGraph, dp, visited, recVisited, longestPath)
-
-		// also not working:
-		if verbose {
-			cur := mainModule
-			// have visited array here too
-			vis := make(map[string]bool)
-			for vis[cur] == false {
-				vis[cur] = true
-				fmt.Print(cur + " -> ")
-				cur = longestPath[cur]
-				// if vis[cur] == true {
-				// 	break
-				// }
-			}
-		}
+		// Get all chains starting from main module
+		chains := make(map[int][]string)
+		var temp []string
+		getChains(mainModule, depGraph, temp, chains)
```

Comment:
would be nicer if getChains returns the chains

Reply:
I also thought of that but got confused a bit. `chains` needs to be kind of like a global variable for the `getChains` function but if I declare it inside `getChains` then I think each time it will get redeclared.

Would using a wrapper function which declares `chains` and then calls `getChains` be ideal? We can then just call this wrapper function in the main code and it will return `chains`.